bitsail-connectors/connector-ftp/src/main/java/com/bytedance/bitsail/connector/ftp/source/split/coordinator/FtpSourceSplitCoordinator.java

```diff
@@ -0,0 +1,176 @@
+    LOG.info("Found reader {}", subtaskId);
+    tryAssignSplitsToReader();
+  }
+
+  @Override
+  public void addSplitsBack(List<FtpSourceSplit> splits, int subtaskId) {
+    LOG.info("Source reader {} return splits {}.", subtaskId, splits);
+
+    int readerNum = context.totalParallelism();
+    for (FtpSourceSplit split : splits) {
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Re-assign split {} to the {}-th reader.", split.uniqSplitId(), readerIndex);
+    }
+
+    tryAssignSplitsToReader();
+  }
+
+  private void tryAssignSplitsToReader() {
+    Map<Integer, List<FtpSourceSplit>> splitsToAssign = new HashMap<>();
+
+    for (Integer readerIndex : splitAssignmentPlan.keySet()) {
+      if (CollectionUtils.isNotEmpty(splitAssignmentPlan.get(readerIndex)) && context.registeredReaders().contains(readerIndex)) {
+        splitsToAssign.put(readerIndex, Lists.newArrayList(splitAssignmentPlan.get(readerIndex)));
+      }
+    }
+
+    for (Integer readerIndex : splitsToAssign.keySet()) {
+      LOG.info("Try assigning splits reader {}, splits are: [{}]", readerIndex,
+              splitsToAssign.get(readerIndex).stream().map(FtpSourceSplit::getSplitId).collect(Collectors.toList()));
+      splitAssignmentPlan.remove(readerIndex);
+      context.assignSplit(readerIndex, splitsToAssign.get(readerIndex));
+      context.signalNoMoreSplits(readerIndex);
+      LOG.info("Finish assigning splits reader {}", readerIndex);
+    }
+  }
+
+  private List<FtpSourceSplit> construct(IFtpHandler ftpHandler) throws IOException {
+    List<FtpSourceSplit> splits = new ArrayList<>();
+    ftpHandler.loginFtpServer();
+    String[] paths = ftpHandler.getFtpConfig().getPaths();
+    int index = 0;
+    for (String path : paths) {
+      if (null != path && path.length() > 0) {
+        path = path.replace("\n", "").replace("\r", "");
+        String[] pathArray = StringUtils.split(path, ",");
+        for (String p : pathArray) {
+          for (String f : ftpHandler.getFiles(p.trim())) {
+            FtpSourceSplit split = new FtpSourceSplit(index++);
+            split.setPath(f);
+            splits.add(split);
+          }
+        }
+      }
+      this.fileSize += ftpHandler.getFilesSize(path);
+    }
+
+    int numSplits = splits.size();
+    ftpHandler.logoutFtpServer();
+    LOG.info("FtpSourceSplitCoordinator Input splits size: {}", numSplits);
+    return splits;
+  }
+
+  @Override
+  public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+      // empty
+  }
+
+  @Override
+  public EmptyState snapshotState() {
+    return new EmptyState();
+  }
+
+  @Override
+  public void close() {
+      // empty
+  }
+
+  @NoArgsConstructor
+  static class ReaderSelector {
+    private static long readerIndex = 0;
+    public static int getReaderIndex(int totalReaderNum) {
```

Comment:
we could add a UT for this I believe

Reply:
I add unit test for this class which named  `testReaderSelector`.